### PR TITLE
Bugfix: notification device should be registered with auth user

### PR DESCRIPTION
**Pull request by alexis-kochetov**
…only

### CODE REVIEW

File: src/dotnet/Api.Contracts/Notification/INotifications.cs [hunk 1]
```diff
@@ -33,6 +35,14 @@ public sealed partial record Notifications_RegisterDevice(
     [property: DataMember, MemoryPackOrder(2)] DeviceType DeviceType
 ) : ISessionCommand<Unit>, IApiCommand;
 
+[DataContract, MemoryPackable(GenerateType.VersionTolerant)]
+// ReSharper disable once InconsistentNaming
+public sealed partial record Notifications_DeregisterDevice(
```
Comment (crui3er) [inline src/dotnet/Api.Contracts/Notification/INotifications.cs:40]: May be, better use 'Unregister. It sounds more familiar. Just like an idea.

File: src/dotnet/App.Maui/Platforms/Android/AndroidDeviceTokenRetriever.cs [hunk 2]
```diff
@@ -22,4 +22,14 @@ public class AndroidDeviceTokenRetriever(IServiceProvider services) : IDeviceTok
             return null;
         }
     }
+
+    public async Task DeleteDeviceToken(CancellationToken cancellationToken)
+    {
+        try {
+            await FirebaseMessaging.Instance.DeleteToken().AsAsync().ConfigureAwait(true);
```
Comment (crui3er) [inline src/dotnet/App.Maui/Platforms/Android/AndroidDeviceTokenRetriever.cs]: Do we need to continueOnCapturedContext?

File: src/dotnet/Notification.Service/Notifications.cs [hunk 3]
```diff
@@ -57,6 +58,21 @@ public class Notifications(IServiceProvider services) : INotifications
         await Commander.Run(registerDeviceCommand, cancellationToken).ConfigureAwait(false);
     }
 
+    // [CommandHandler]
+    public virtual async Task OnDeregisterDevice(
+        Notifications_DeregisterDevice command, CancellationToken cancellationToken)
+    {
+        var (session, deviceId) = command;
+        var account = await Accounts.GetOwn(session, cancellationToken).ConfigureAwait(false);
+        var existingDevices = await Backend.ListDevices(account.Id, cancellationToken).ConfigureAwait(false);
```
Comment (crui3er) [inline src/dotnet/Notification.Service/Notifications.cs:67]: I think that deviceId is primary key for device and we should get device by id and then remove it if it exists.
Also we can log warn if existent device userId is different from the account id.

File: src/dotnet/Notification.UI.Blazor/Module/NotificationBlazorUIModule.cs [hunk 4]
```diff
@@ -15,6 +16,7 @@ protected override void InjectServices(IServiceCollection services)
     {
         services.AddFusion();
         services.AddScoped<NotificationUI>();
+        services.AddScoped<INotificationUI, NotificationUI>(c => c.GetRequiredService<NotificationUI>());
```
Comment (crui3er) [inline src/dotnet/Notification.UI.Blazor/Module/NotificationBlazorUIModule.cs]: We have more short syntax for the same:
services.AddAlias<INotificationUI, NotificationUI>(ServiceLifetime.Scoped)

File: src/dotnet/Notification.UI.Blazor/NotificationUI.cs [hunk 5]
```diff
@@ -132,29 +132,70 @@ public void SetPermissionState(string permissionState)
...
+    public async Task EnsureDeviceRegistered(CancellationToken cancellationToken = default)
+    {
+        var deviceId = await DeviceTokenRetriever.GetDeviceToken(cancellationToken).ConfigureAwait(false);
+        var existingTask = _registerDeviceTask;
+        if (existingTask != null) {
+            var alreadyRegisteredDeviceId = await existingTask.ConfigureAwait(false);
+            if (alreadyRegisteredDeviceId == deviceId)
+                return;
+        }
+        _registerDeviceTask = null;
```
Comment (crui3er) [inline src/dotnet/Notification.UI.Blazor/NotificationUI.cs]: you need to synchronize write access to _registerDeviceTask field

File: src/dotnet/UI.Blazor/Services/AccountUI/AccountUI.cs [hunk 6]
```diff
@@ -61,6 +63,20 @@ public TimeSpan GetPostChangeInvalidationDelay(TimeSpan maxInvalidationDelay)
         return (changedAt + maxInvalidationDelay - CpuClock.Now).Positive();
     }
 
-    public Task SignOut()
-        => ClientAuth.SignOut();
+    public async Task SignIn(string schema)
+    {
+        await ClientAuth.SignIn(schema).ConfigureAwait(false);
```
Comment (crui3er) [inline src/dotnet/UI.Blazor/Services/AccountUI/AccountUI.cs:68]: ClientAuth.SignIn is used on ProviderSelectStep. Don't we need to replace it with AccountUI.SignIn?

**Review by crui3er — CHANGES_REQUESTED** [review]

File: src/dotnet/Api.Contracts/Notification/INotifications.cs (hunk 1 above)
Comment (alexis-kochetov) [reply src/dotnet/Api.Contracts/Notification/INotifications.cs:40]: Unregister has different meaning - https://english.stackexchange.com/questions/25931/unregister-vs-deregister

File: src/dotnet/App.Maui/Platforms/Android/AndroidDeviceTokenRetriever.cs (hunk 2 above)
Comment (alexis-kochetov) [reply src/dotnet/App.Maui/Platforms/Android/AndroidDeviceTokenRetriever.cs]: AI code completion is a mess!

File: src/dotnet/Notification.Service/Notifications.cs (hunk 3 above)
Comment (alexis-kochetov) [reply src/dotnet/Notification.Service/Notifications.cs:67]: I think we should be able to deregister only own device Ids - that's why I'm checking the list of existing devices for the user.

File: src/dotnet/Notification.UI.Blazor/NotificationUI.cs (hunk 5 above)
Comment (alexis-kochetov) [reply src/dotnet/Notification.UI.Blazor/NotificationUI.cs]: OK. it will be simpler

**Pull request merged**